File: api_jogos_ia/loteria_ai_engine/app/core/genetic_engine.py
"""
Genetic Algorithm Core Implementation for Loteria AI Engine
"""
import numpy as np
import random
from typing import List, Dict, Tuple, Any
from deap import base, creator, tools, algorithms
import json


class LotteryGeneticEngine:
    """
    Genetic Algorithm engine for evolving lottery seeds.
    Each individual represents a "seed" with weights and correlation matrices.
    """

    def __init__(self, game_type: str, num_numbers: int, numbers_to_pick: int):
        self.game_type = game_type
        self.num_numbers = num_numbers  # e.g., 25 for Lotofacil
        self.numbers_to_pick = numbers_to_pick  # e.g., 15 for Lotofacil
        
        # Setup DEAP framework
        self._setup_deap()
        
        # Historical data for fitness evaluation
        self.historical_results: List[List[int]] = []
        
        # Current best seed
        self.best_individual = None
        self.best_fitness = 0.0
        self.current_generation = 0
        
        # Evolution statistics for visualization
        self.generation_history: List[Dict[str, Any]] = []

    def _setup_deap(self):
        """Setup DEAP creator and toolbox"""
        # Create fitness and individual classes
        try:
            creator.create("FitnessMax", base.Fitness, weights=(1.0,))
            creator.create("Individual", list, fitness=creator.FitnessMax)
        except AttributeError:
            # Already created
            pass
        
        self.toolbox = base.Toolbox()
        
        # Attribute generator: weights for each number (0.0 to 1.0)
        self.toolbox.register("attr_float", random.random)
        
        # Structure initializers
        self.toolbox.register(
            "individual",
            tools.initRepeat,
            creator.Individual,
            self.toolbox.attr_float,
            self.num_numbers
        )
        
        # Population initializer
        self.toolbox.register(
            "population",
            tools.initRepeat,
            list,
            self.toolbox.individual
        )
        
        # Operators
        self.toolbox.register("evaluate", self._fitness_function)
        self.toolbox.register("mate", tools.cxBlend, alpha=0.5)
        self.toolbox.register("mutate", tools.mutGaussian, mu=0, sigma=0.1, indpb=0.1)
        self.toolbox.register("select", tools.selTournament, tournsize=3)

    def load_historical_data(self, results: List[List[int]]):
        """Load historical results for fitness evaluation"""
        self.historical_results = results

    def _fitness_function(self, individual: List[float]) -> Tuple[float]:
        """
        Evaluate the fitness of an individual (seed).
        Simulates games and compares against historical results.
        """
        weights = np.array(individual)
        weights = weights / weights.sum()  # Normalize
        
        # Generate correlation matrix based on individual
        correlation_matrix = self._generate_correlation_matrix(individual)
        
        # Simulate games
        simulated_games = self._simulate_games(weights, correlation_matrix, n_games=1000)
        
        # Score against historical results
        score = self._score_against_history(simulated_games)
        
        return (score,)

    def _generate_correlation_matrix(self, individual: List[float]) -> np.ndarray:
        """Generate a correlation matrix based on the individual's genome"""
        # Simplified: create correlations based on number proximity and weights
        matrix = np.zeros((self.num_numbers, self.num_numbers))
        for i in range(self.num_numbers):
            for j in range(self.num_numbers):
                if i != j:
                    # Higher correlation for nearby numbers
                    distance = abs(i - j)
                    base_corr = max(0, 1.0 - distance / 10.0)
                    weight_factor = (individual[i] + individual[j]) / 2.0
                    matrix[i, j] = base_corr * weight_factor
        return matrix

    def _simulate_games(self, weights: np.ndarray, 
                       correlation_matrix: np.ndarray, 
                       n_games: int = 1000) -> List[List[int]]:
        """Simulate lottery games based on weights and correlations"""
        games = []
        for _ in range(n_games):
            # Adjust probabilities based on correlations (simplified)
            probs = weights.copy()
            
            # Select numbers without replacement
            selected = []
            available_probs = probs.copy()
            
            for _ in range(self.numbers_to_pick):
                # Normalize remaining probabilities
                available_probs = available_probs / available_probs.sum()
                
                # Select number based on probability
                chosen = np.random.choice(
                    self.num_numbers, 
                    p=available_probs
                )
                selected.append(chosen + 1)  # 1-indexed
                
                # Remove chosen number
                available_probs[chosen] = 0
            
            games.append(sorted(selected))
        
        return games

    def _score_against_history(self, simulated_games: List[List[int]]) -> float:
        """
        Score simulated games against historical results.
        Rewards matches of partial results (quadras, quinas, etc.)
        """
        if not self.historical_results:
            return 0.0
        
        total_score = 0.0
        n_comparisons = min(len(simulated_games), len(self.historical_results) * 10)
        
        for sim_game in simulated_games[:n_comparisons]:
            sim_set = set(sim_game)
            
            # Compare with multiple historical results
            for hist_result in self.historical_results[-100:]:  # Last 100 results
                hist_set = set(hist_result)
                matches = len(sim_set.intersection(hist_set))
                
                # Scoring: exponential reward for more matches
                if matches >= self.numbers_to_pick - 2:  # Near win
                    total_score += 1000.0
                elif matches >= self.numbers_to_pick - 3:
                    total_score += 100.0
                elif matches >= self.numbers_to_pick - 4:
                    total_score += 10.0
                elif matches >= self.numbers_to_pick // 2:
                    total_score += 1.0
        
        return total_score / n_comparisons

    def evolve(self, population_size: int = 100, generations: int = 50, 
               checkpoint_callback=None) -> Dict[str, Any]:
        """
        Run the evolutionary algorithm.
        Returns the best seed found.
        """
        population = self.toolbox.population(n=population_size)
        
        # Statistics setup
        stats = tools.Statistics(lambda ind: ind.fitness.values[0])
        stats.register("avg", np.mean)
        stats.register("max", np.max)
        
        # Hall of Fame
        hof = tools.HallOfFame(1)
        
        # Evolve
        population, logbook = algorithms.eaSimple(
            population,
            self.toolbox,
            cxpb=0.7,
            mutpb=0.2,
            ngen=generations,
            stats=stats,
            halloffame=hof,
            verbose=False
        )
        
        # Update best
        self.best_individual = hof[0]
        self.best_fitness = hof[0].fitness.values[0]
        self.current_generation += generations
        
        # Record history for visualization
        for record in logbook:
            self.generation_history.append({
                "generation": self.current_generation - generations + record['gen'],
                "best_fitness": record['max'],
                "avg_fitness": record['avg'],
                "timestamp": datetime.now()
            })
        
        # Callback for checkpointing
        if checkpoint_callback:
            checkpoint_callback(self.get_seed_data())
        
        return self.get_seed_data()

    def get_seed_data(self) -> Dict[str, Any]:
        """Convert current best individual to seed data format"""
        if self.best_individual is None:
            # Return random seed if no evolution yet
            weights = [random.random() for _ in range(self.num_numbers)]
        else:
            weights = list(self.best_individual)
        
        # Normalize weights
        weights = [w / sum(weights) for w in weights]
        
        return {
            "version": f"1.0.{self.current_generation}",
            "game_type": self.game_type,
            "weights": weights,
            "correlations": self._generate_correlation_matrix(
                self.best_individual if self.best_individual else weights
            ).tolist() if self.best_individual else None,
            "filters": {
                "min_evens": self.numbers_to_pick // 2 - 2,
                "max_evens": self.numbers_to_pick // 2 + 2,
                "min_sum": self.numbers_to_pick * (self.num_numbers // 4),
                "max_sum": self.numbers_to_pick * (self.num_numbers * 3 // 4)
            },
            "generation": self.current_generation,
            "fitness_score": self.best_fitness
        }

    def generate_game_from_seed(self, seed_data: Dict[str, Any]) -> List[int]:
        """Generate a single game using the seed data"""
        weights = np.array(seed_data["weights"])
        
        # Select numbers based on weights
        selected = []
        available_probs = weights.copy()
        
        for _ in range(self.numbers_to_pick):
            available_probs = available_probs / available_probs.sum()
            chosen = np.random.choice(self.num_numbers, p=available_probs)
            selected.append(chosen + 1)  # 1-indexed
            available_probs[chosen] = 0
        
        return sorted(selected)


# Import datetime for generation history
from datetime import datetime

File: web/lib/genetic_engine.js
/**
 * Genetic Algorithm Engine for Lottery Number Generation
 * 
 * Ported from Python DEAP implementation to pure JavaScript.
 * No external dependencies required — runs inside Node.js.
 * 
 * Each "individual" is an array of 25 weights (0.0–1.0) representing
 * the probability bias for each Lotofácil number (01–25).
 * The algorithm evolves these weights to maximize matches against
 * historical results.
 */

const fs = require('fs');
const path = require('path');

// Janela de concursos usada na avaliação de fitness (os N mais recentes).
// Trade-off: janela maior = aprendizado mais amplo, mas evolução mais lenta.
// Medido (100pop x 20gen): 100→~4min · 300→~6min · 500→~9.6min · 3750→~55min.
// Configurável via env FITNESS_WINDOW_SIZE (ex.: 100 para evolução rápida,
// 500+ para aprendizado mais amplo). Default 300 = bom equilíbrio.
const FITNESS_WINDOW_SIZE = parseInt(process.env.FITNESS_WINDOW_SIZE || '300', 10) || 300;

class LotteryGeneticEngine {
  constructor(gameType = 'LOTOFACIL', options = {}) {
    this.gameType = gameType;
    this.numNumbers = 25;       // Lotofácil: 25 numbers
    this.numbersToPick = 15;    // Lotofácil: pick 15

    this.historicalResults = []; // [[n1,n2,...], ...] — loaded from cache
    this.historicalSets = [];   // Pre-computed Sets for fast fitness evaluation
    this.currentSeed = null;     // Best weights found so far
    this.currentGeneration = 0;
    this.bestFitness = 0;
    this.generationHistory = [];

    this.dbPath = path.join(__dirname, '..', 'database', 'lotofacil.json');
    this.seedsPath = path.join(__dirname, '..', 'database', 'seeds.json');

    // Fontes de dados injetáveis (para ambientes serverless como Vercel)
    this.historicalResultsProvider = options.historicalResultsProvider || null;
    this.seedProvider = options.seedProvider || null;
    this.seedSaver = options.seedSaver || null;
    this.autoEvolve = options.autoEvolve !== undefined ? options.autoEvolve : true;
  }

  // ======================================================================
  //  PERSISTENCE
  // ======================================================================

  /** Load historical results from the local JSON cache */
  loadHistoricalResults() {
    if (this.historicalResultsProvider) {
      const data = this.historicalResultsProvider();
      if (data && data.length) {
        this.historicalResults = data
          .filter(c => c.listaDezenas && Array.isArray(c.listaDezenas))
          .map(c => c.listaDezenas.map(n => parseInt(n)));
        this.historicalSets = this.historicalResults.slice(-FITNESS_WINDOW_SIZE).map(r => new Set(r));
        console.log(`🧬 Engine: ${this.historicalResults.length} resultados históricos carregados (DB)`);
        return true;
      }
      return false;
    }
    try {
      if (fs.existsSync(this.dbPath)) {
        const raw = fs.readFileSync(this.dbPath, 'utf8');
        const data = JSON.parse(raw);
        this.historicalResults = data
          .filter(c => c.listaDezenas && Array.isArray(c.listaDezenas))
          .map(c => c.listaDezenas.map(n => parseInt(n)));
        // Pre-compute Sets for fast matching — evita recriar 75M de Sets na evolução
        this.historicalSets = this.historicalResults.slice(-FITNESS_WINDOW_SIZE).map(r => new Set(r));
        console.log(`🧬 Engine: ${this.historicalResults.length} resultados históricos carregados`);
        return true;
      }
    } catch (e) {
      console.error('Erro ao carregar histórico:', e.message);
    }
    return false;
  }

  /** Load the best seed from disk (if one was saved previously) */
  loadSavedSeed() {
    if (this.seedProvider) {
      const data = this.seedProvider();
      if (data && data.weights) {
        this.currentSeed = data.weights;
        this.currentGeneration = data.generation || 0;
        this.bestFitness = data.fitness || 0;
        this.generationHistory = data.history || [];
        console.log(`🧬 Engine: semente carregada (DB) — geração ${this.currentGeneration}, fitness ${this.bestFitness.toFixed(4)}`);
        return true;
      }
      return false;
    }
    try {
      if (fs.existsSync(this.seedsPath)) {
        const raw = fs.readFileSync(this.seedsPath, 'utf8');
        const data = JSON.parse(raw);
        this.currentSeed = data.weights;
        this.currentGeneration = data.generation || 0;
        this.bestFitness = data.fitness || 0;
        this.generationHistory = data.history || [];
        console.log(`🧬 Engine: semente carregada — geração ${this.currentGeneration}, fitness ${this.bestFitness.toFixed(4)}`);
        return true;
      }
    } catch (e) {
      console.error('Erro ao carregar semente:', e.message);
    }
    return false;
  }

  /** Persist the current best seed to disk */
  saveSeed() {
    if (this.seedSaver) {
      const data = {
        version: `1.0.${this.currentGeneration}`,
        game_type: this.gameType,
        weights: this.currentSeed,
        generation: this.currentGeneration,
        fitness: this.bestFitness,
        history: this.generationHistory.slice(-500),
        updatedAt: new Date().toISOString()
      };
      try { this.seedSaver(data); } catch (e) { console.error('Erro ao salvar semente (DB):', e.message); }
      return;
    }
    try {
      const dir = path.dirname(this.seedsPath);
      if (!fs.existsSync(dir)) fs.mkdirSync(dir, { recursive: true });
      const data = {
        version: `1.0.${this.currentGeneration}`,
        game_type: this.gameType,
        weights: this.currentSeed,
        generation: this.currentGeneration,
        fitness: this.bestFitness,
        history: this.generationHistory.slice(-500),
        updatedAt: new Date().toISOString()
      };
      fs.writeFileSync(this.seedsPath, JSON.stringify(data, null, 2), 'utf8');
    } catch (e) {
      console.error('Erro ao salvar semente:', e.message);
    }
  }

  // ======================================================================
  //  GENETIC ALGORITHM — CORE OPERATIONS
  // ======================================================================

  /** Create a random individual: 25 weights in [0, 1] */
  createIndividual() {
    const ind = [];
    for (let i = 0; i < this.numNumbers; i++) ind.push(Math.random());
    return ind;
  }

  /** Normalise an array so that all elements sum to 1 */
  normalise(weights) {
    const sum = weights.reduce((a, b) => a + b, 0);
    if (sum <= 0) return weights.map(() => 1 / weights.length);
    return weights.map(w => w / sum);
  }

  /**
   * Simulate N lottery games given a set of weights.
   * Uses roulette-wheel selection without replacement.
   * @param {number|null} pickCount — nº de dezenas por jogo (ex.: 15–20 na
   *   Lotofácil). Se omitido, usa o padrão da loteria (numbersToPick).
   */
  simulateGames(weights, nGames = 500, pickCount = null) {
    const probs = this.normalise(weights);
    const games = [];
    const target = pickCount || this.numbersToPick;

    for (let g = 0; g < nGames; g++) {
      const selected = [];
      const available = [...probs];

      for (let p = 0; p < target; p++) {
        const total = available.reduce((a, b) => a + b, 0);
        const normed = available.map(w => w / total);

        let r = Math.random();
        let chosen = 0;
        for (let i = 0; i < this.numNumbers; i++) {
          r -= normed[i];
          if (r <= 0) { chosen = i; break; }
        }

        selected.push(chosen + 1); // 1-indexed
        available[chosen] = 0;     // remove from pool
      }

      games.push(selected.sort((a, b) => a - b));
    }

    return games;
  }

  /**
   * Fitness function.
   * Simulates games with the individual's weights and compares them
   * against the last 100 historical results. Rewards games that
   * produce many partial matches (11, 12, 13, 14, 15 hits).
   */
  evaluateFitness(individual) {
    const simulated = this.simulateGames(individual, 500);
    if (!this.historicalSets.length) return 0;

    let totalScore = 0;

    for (const simGame of simulated) {
      for (const histSet of this.historicalSets) {
        let matches = 0;
        for (const n of simGame) if (histSet.has(n)) matches++;

        if (matches >= 15)      totalScore += 10000;
        else if (matches >= 14) totalScore += 1000;
        else if (matches >= 13) totalScore += 100;
        else if (matches >= 12) totalScore += 10;
        else if (matches >= 11) totalScore += 1;
      }
    }

    return totalScore / simulated.length;
  }

  /** Tournament selection: pick the best from `tournamentSize` random individuals */
  tournamentSelect(population, fitnesses, tournamentSize = 3) {
    let bestIdx = -1;
    let bestFit = -Infinity;
    for (let i = 0; i < tournamentSize; i++) {
      const idx = Math.floor(Math.random() * population.length);
      if (fitnesses[idx] > bestFit) { bestFit = fitnesses[idx]; bestIdx = idx; }
    }
    return population[bestIdx];
  }

  /**
   * Blend crossover (DEAP's cxBlend).
   * Children are placed in an interval [min-ext, max+ext] between the parents.
   */
  crossover(p1, p2, alpha = 0.5) {
    const c1 = []; const c2 = [];
    for (let i = 0; i < this.numNumbers; i++) {
      const lo = Math.min(p1[i], p2[i]);
      const hi = Math.max(p1[i], p2[i]);
      const ext = (hi - lo) * alpha;
      const low = lo - ext;
      const high = hi + ext;
      c1[i] = Math.max(0, Math.min(1, low + Math.random() * (high - low)));
      c2[i] = Math.max(0, Math.min(1, low + Math.random() * (high - low)));
    }
    return [c1, c2];
  }

  /**
   * Gaussian mutation (DEAP's mutGaussian).
   * Each gene has `indpb` probability of being perturbed by N(0, sigma).
   */
  mutate(individual, sigma = 0.1, indpb = 0.1) {
    const m = [...individual];
    for (let i = 0; i < m.length; i++) {
      if (Math.random() < indpb) {
        // Box-Muller transform → standard normal
        const u1 = Math.random();
        const u2 = Math.random();
        const z = Math.sqrt(-2 * Math.log(u1)) * Math.cos(2 * Math.PI * u2);
        m[i] = Math.max(0, Math.min(1, m[i] + z * sigma));
      }
    }
    return m;
  }

  // ======================================================================
  //  EVOLUTION LOOP
  // ======================================================================

  /**
   * Run the evolutionary algorithm.
   * @param {number} populationSize   — default 100
   * @param {number} generations      — default 20 (lighter for first run)
   * @returns {object|null} seed data
   */
  evolve(populationSize = 100, generations = 20) {
    if (!this.historicalResults.length) {
      console.warn('⚠️ Sem dados históricos — evolução cancelada');
      return null;
    }

    console.log(`🧬 Evolução: ${generations} gerações, ${populationSize} indivíduos`);
    const t0 = Date.now();

    // 1. Initialise population
    let population = [];
    for (let i = 0; i < populationSize; i++) population.push(this.createIndividual());

    let bestInd = null;
    let bestFit = -Infinity;

    for (let gen = 0; gen < generations; gen++) {
      // 2. Evaluate
      const fitnesses = population.map(ind => this.evaluateFitness(ind));

      const genBest = Math.max(...fitnesses);
      const genAvg = fitnesses.reduce((a, b) => a + b, 0) / fitnesses.length;
      const bestIdx = fitnesses.indexOf(genBest);

      if (genBest > bestFit) { bestFit = genBest; bestInd = [...population[bestIdx]]; }

      // 3. Build next generation (elitism: keep top 2)
      const ranked = population
        .map((ind, i) => ({ ind, fit: fitnesses[i] }))
        .sort((a, b) => b.fit - a.fit);

      const next = [];
      next.push([...ranked[0].ind]);
      next.push([...ranked[1].ind]);

      while (next.length < populationSize) {
        const p1 = this.tournamentSelect(population, fitnesses);
        const p2 = this.tournamentSelect(population, fitnesses);
        let [c1, c2] = [p1, p2];

        if (Math.random() < 0.7) [c1, c2] = this.crossover(p1, p2);
        if (Math.random() < 0.2) c1 = this.mutate(c1);
        if (Math.random() < 0.2) c2 = this.mutate(c2);

        next.push(c1);
        if (next.length < populationSize) next.push(c2);
      }

      population = next;

      // 4. Log progress every 10 generations
      if ((gen + 1) % 10 === 0 || gen === 0) {
        console.log(`   Gen ${gen + 1}/${generations} | melhor: ${genBest.toFixed(4)} | média: ${genAvg.toFixed(4)}`);
      }

      this.generationHistory.push({
        generation: this.currentGeneration + gen + 1,
        bestFitness: genBest,
        avgFitness: genAvg,
        timestamp: new Date().toISOString()
      });
    }

    this.currentSeed = bestInd;
    this.currentGeneration += generations;
    this.bestFitness = bestFit;
    this.saveSeed();

    const elapsed = ((Date.now() - t0) / 1000).toFixed(1);
    console.log(`✅ Evolução concluída em ${elapsed}s · fitness: ${bestFit.toFixed(4)}`);

    return this.getSeed();
  }

  // ======================================================================
  //  PUBLIC API
  // ======================================================================

  /** Return the current best seed metadata */
  getSeed() {
    if (!this.currentSeed) this.currentSeed = this.createIndividual();
    return {
      version: `1.0.${this.currentGeneration}`,
      game_type: this.gameType,
      weights: this.normalise(this.currentSeed),
      generation: this.currentGeneration,
      fitnessScore: this.bestFitness,
      status: this.currentGeneration > 0 ? 'evolved' : 'initial'
    };
  }

  /**
   * Generate `quantity` games using the current best seed.
   * @param {number} quantity   — quantos jogos gerar
   * @param {number|null} pickCount — quantas dezenas por jogo (15–20 na
   *   Lotofácil, respeitando a tabela da Caixa). Omitido = padrão da loteria.
   */
  generateGames(quantity = 5, pickCount = null) {
    if (this.autoEvolve && !this.currentSeed && this.historicalResults.length > 0) {
      this.evolve(100, 15); // quick first evolution
    }
    const seed = this.currentSeed || this.createIndividual();
    const games = this.simulateGames(seed, quantity, pickCount);
    return {
      game_type: this.gameType,
      seed_version: `1.0.${this.currentGeneration}`,
      pickCount: pickCount || this.numbersToPick,
      games,
      generatedAt: new Date().toISOString()
    };
  }

  /**
   * Return the full evolution history for the dashboard
   */
  getEvolutionHistory() {
    return {
      gameType: this.gameType,
      currentGeneration: this.currentGeneration,
      bestFitness: this.bestFitness,
      populationSize: 100,
      history: this.generationHistory.slice(-200), // últimas 200 gerações
      seedInfo: {
        version: `1.0.${this.currentGeneration}`,
        weights: this.currentSeed ? this.normalise(this.currentSeed) : null,
        fitnessScore: this.bestFitness
      },
      stats: {
        totalGenerations: this.currentGeneration,
        totalHistoryPoints: this.generationHistory.length,
        contestsAnalyzed: this.historicalResults.length,
        evolutionTime: this.currentGeneration * 2.5, // ~2.5s por geração
        lastUpdate: new Date().toISOString()
      }
    };
  }

  /**
   * Continue evolving for N more generations.
   * Non-blocking: runs via setImmediate to not freeze the server.
   * Returns immediately with a promise that resolves when done.
   */
  async evolveMore(generations = 10) {
    if (this._evolving) {
      throw new Error('Evolução já em andamento');
    }
    this._evolving = true;
    try {
      return await new Promise((resolve) => {
        // Usa setImmediate para não travar o event loop completamente
        // A cada 5 gerações, dá um breather pro servidor respirar
        const totalGens = generations;
        let completed = 0;

        const runBatch = () => {
          setImmediate(() => {
            const batchSize = Math.min(5, totalGens - completed);
            this.evolve(100, batchSize);
            completed += batchSize;
            if (completed < totalGens) {
              runBatch();
            } else {
              this._evolving = false;
              resolve(this.getEvolutionHistory());
            }
          });
        };

        runBatch();
      });
    } catch (e) {
      this._evolving = false;
      throw e;
    }
  }

  /** Check if evolution is currently running */
  isEvolving() {
    return !!this._evolving;
  }

  /**
   * Bootstrap the engine:
   *  - Load historical results from local cache
   *  - Load saved seed (if any)
   *  - If no seed exists, run a quick evolution
   */
  initialize() {
    const hasHistory = this.loadHistoricalResults();
    const hasSeed = this.loadSavedSeed();

    if (this.autoEvolve && hasHistory && !hasSeed) {
      console.log('🧬 Nenhuma semente salva. Executando evolução inicial...');
      this.evolve(100, 15);
    }

    return {
      ready: hasHistory || hasSeed,
      contestsLoaded: this.historicalResults.length,
      generation: this.currentGeneration,
      fitness: this.bestFitness
    };
  }
}

module.exports = LotteryGeneticEngine;
